tests/unit/metrics/test_ffwd.py

```diff
@@ -0,0 +1,208 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numbers
+
+import pytest
+
+from gordon.metrics import ffwd
+
+
+def test_ffwd_protocol_connection_made(mocker):
+    """Message sends and connection closes."""
+    mock_transport = mocker.Mock()
+    message = 'Testing message!'
+
+    protocol = ffwd.FfwdClientProtocol(message)
+    protocol.connection_made(mock_transport)
+
+    mock_transport.sendto.assert_called_once_with(message)
+    mock_transport.close.assert_called_once_with()
+
+
+@pytest.mark.asyncio
+async def test_udp_client_send(mocker):
+    """Sends serialized metric using FfwdClientProtocol."""
+    mock_loop = mocker.Mock()
+
+    actual_func = None
+    actual_addr = None
+
+    async def mock_create_datagram_endpoint(func, remote_addr):
+        nonlocal actual_func
+        nonlocal actual_addr
+        actual_func = func
+        actual_addr = remote_addr
+
+    mock_loop.create_datagram_endpoint = mock_create_datagram_endpoint
+
+    udp_client = ffwd.UDPClient('1.2.3.4', '5678', mock_loop)
+    await udp_client.send({'an': 'object'})
+
+    actual_protocol = actual_func()
+    assert isinstance(actual_protocol, ffwd.FfwdClientProtocol)
+    assert ('1.2.3.4', '5678') == actual_addr
+    assert b'{"an": "object"}' == actual_protocol.message
+
+
+@pytest.fixture
+def mock_udp_client_and_metrics_sent(mocker):
+    metrics_sent = []
+
+    async def mock_send(metric):
+        nonlocal metrics_sent
+        metrics_sent.append(metric)
+
+    mock_udp_client = mocker.Mock()
+    mock_udp_client.send = mock_send
+    return mock_udp_client, metrics_sent
+
+
+@pytest.fixture
+def timer_and_metrics_sent(mock_udp_client_and_metrics_sent):
+    mock_udp_client, metrics_sent = mock_udp_client_and_metrics_sent
+    metric = {'a': 'metric', 'value': None}
+    timer = ffwd.FfwdTimer(metric, mock_udp_client)
+    return timer, metrics_sent
+
+
+@pytest.mark.asyncio
+async def test_ffwdtimer_manual(timer_and_metrics_sent):
+    """Times when operated manually."""
+    timer, metrics_sent = timer_and_metrics_sent
+    await timer.start()
+    assert isinstance(timer._start_time, numbers.Number)
+
+    await timer.stop()
+    assert len(metrics_sent) == 1
+    assert 'metric' == metrics_sent[0]['a']
+    assert 0 < metrics_sent[0]['value']
+
+
+@pytest.mark.asyncio
+async def test_ffwdtimer_as_context_manager(timer_and_metrics_sent):
+    """Times as a context manager."""
+    timer, metrics_sent = timer_and_metrics_sent
+    async with timer as t:
+        assert isinstance(t._start_time, numbers.Number)
+
+    assert len(metrics_sent) == 1
+    assert 'metric' == metrics_sent[0]['a']
+    assert 0 < metrics_sent[0]['value']
+
+
+@pytest.fixture
+def test_config(mocker, monkeypatch):
```

Comment:
The name should indicate that it also has a mock.  Or do the mock separately.

Reply:
Hmm I don't even remember including the monkeypatch. I don't think it's used. I'll remove it.